src/util/psbt/mod.rs

```diff
@@ -196,6 +206,520 @@ impl PartiallySignedTransaction {
 
         Ok(())
     }
+
+    /// Attempts to create _all_ the required signatures for this PSBT using `k`.
+    ///
+    /// **NOTE**: Taproot inputs are, as yet, not supported by this function. We currently only
+    /// attempt to sign ECDSA inputs.
+    ///
+    /// If you just want to sign an input with one specific key consider using `sighash_ecdsa`. This
+    /// function does not support scripts that contain `OP_CODESEPARATOR`.
+    ///
+    /// # Returns
+    ///
+    /// Either Ok(SigningKeys) or Err((SigningKeys, SigningErrors)), where
+    /// - SigningKeys: A map of input index -> pubkey associated with secret key used to sign.
+    /// - SigningKeys: A map of input index -> the error encountered while attempting to sign.
+    ///
+    /// If an error is returned some signatures may already have been added to the PSBT. Since
+    /// `partial_sigs` is a [`BTreeMap`] it is safe to retry, previous sigs will be overwritten.
+    pub fn sign<C: Signing>(
+        &mut self,
+        k: impl GetKey + Copy,
+        secp: &Secp256k1<C>,
+    ) -> Result<SigningKeys, (SigningKeys, SigningErrors)> {
+        let tx = self.unsigned_tx.clone(); // clone because we need to mutably borrow when signing.
+        let mut cache = SighashCache::new(&tx);
+
+        let mut used = BTreeMap::new();
+        let mut errors = BTreeMap::new();
+
+        for i in 0..self.inputs.len() {
+            match self.signing_algorithm(i).expect("i is valid") {
```

Comment:
I agree your suggested code is also correct but I don't understand why the original code is incorrect. `signing_algorithm` only returns either `Ecdsa` or `Schnorr` and we `continue` if its `Schnorr`?

Reply:
It would panic on any input containing segwitv2. The expect statement is not checking the bounds, but also that the output type is recognizable.

Reply:
Towards the very end of `output_type` definition, you would see another error condition.

Reply:
I don't know what segwitv2 is? I thought we only had v0 and v1 (taproot). Are you talking about future proofing?

Reply:
Yes, this is about future-proofing. Let's say the there is a segwitv2 upgrade. And the current wallet is trying a psbt where one input is segwitv0(ecdsa_sig) and another is segwitv2(unknown type). The current code would crash for no reason because it fails to recognize an output it does not even own.

Reply:
Awesome, cheers. Will use as suggested. Thanks for being patient with me!

Reply:
Added a test case here and another awkward thing I found while trying to use the API. Needed to remove the `Copy` bound and use static dispatch here.

Reply:
https://github.com/sanket1729/rust-bitcoin/commit/be988d5b4dc73693944863dcb9abd5b05ea6dbb2

Reply:
I added your test, had to debug it and I discovered that currently we have to have the `bip32_derivation` field set for an input to be able to sign it, even when using `KeySource::Pubkey` (which does not use keysource). Is this correct? Does a PSBT always have this field set for an input?